llama_index/indices/query/query_transform/feedback_transform.py

```diff
@@ -3,29 +3,65 @@
 from llama_index.evaluation.base import Evaluation
 from llama_index.indices.query.query_transform.base import BaseQueryTransform
 from llama_index.indices.query.schema import QueryBundle
+from llama_index.langchain_helpers.chain_wrapper import LLMPredictor
+from llama_index.llm_predictor.base import BaseLLMPredictor
+from llama_index.prompts.base import Prompt
+from llama_index.response.schema import Response
+
+DEFAULT_RESYNTHESIS_PROMPT_TMPL = (
+    "Here is the original query:\n"
+    "{query_str}\n"
+    "Here is the response given:\n"
+    "{response}\n"
+    "Here is some feedback from evaluator about the response given.\n"
+    "{feedback}\n"
+    "If you want to resynthesize the query, please return the modified query below.\n"
+    "Otherwise, please return the original query.\n"
+)
+
+DEFAULT_RESYNTHESIS_PROMPT = Prompt(DEFAULT_RESYNTHESIS_PROMPT_TMPL)
 
 
 class FeedbackQueryTransformation(BaseQueryTransform):
     """Transform the query given the evaluation feedback.
 
     Args:
         eval(Evaluation): An evaluation object.
+        llm_predictor(BaseLLMPredictor): An LLM predictor.
+        resynthesize_query(bool): Whether to resynthesize the query.
+        resynthesis_prompt(Prompt): A prompt for resynthesizing the query.
 
     """
 
     def __init__(
         self,
         evaluation: Evaluation,
+        llm_predictor: Optional[BaseLLMPredictor] = None,
+        resynthesize_query: bool = False,
+        resynthesis_prompt: Optional[Prompt] = None,
     ) -> None:
         super().__init__()
         self.evaluation = evaluation
+        self.llm_predictor = llm_predictor or LLMPredictor()
+        self.should_resynthesize_query = resynthesize_query
+        self.resynthesis_prompt = resynthesis_prompt or DEFAULT_RESYNTHESIS_PROMPT
 
     def _run(self, query_bundle: QueryBundle, extra_info: Dict) -> QueryBundle:
         query_str = query_bundle.query_str
+        if self.should_resynthesize_query:
+            resynthesized_query = self.resynthesize_query(
+                query_str, self.evaluation.response, self.evaluation.feedback
```

Comment:
it would be great to log the intermediate steps (both via logger, as well as `print_text` with a `verbose` toggle in the `__init__`)

Reply:
e.g. log the initial response, log the new query and/or the feedback.

We should also consider using the callback manager here

Reply:
I have a logger debug in the resynthesize_query function. Also I don't believe by default BaseQueryTransform has a callback manager.

Reply:
Should we add a ticket to hook callback manager into the query transform objects?

I think if we consider query transformation to be a fundamental building block in llama_index, it makes sense to properly trace it as its own even type.

Reply:
I added an issue to the linear backlog